gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+#include <gcs/constraints/smart_table.hh>
+#include <gcs/innards/propagators.hh>
+
+#include <algorithm>
+#include <map>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include <gcs/exception.hh>
+#include <util/overloaded.hh>
+
+using std::make_unique;
+using std::move;
+using std::pair;
+using std::set;
+using std::tuple;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+using std::visit;
+
+using namespace gcs;
+using namespace gcs::innards;
+
+SmartTable::SmartTable(const vector<IntegerVariableID> & v, SmartTuples & t) :
+    _vars(v),
+    _tuples(t)
+{
+}
+
+namespace
+{
+    // Shorthands
+    using VariableDomainMap = unordered_map<IntegerVariableID, vector<Integer>>;
```

Comment:
Not sure... I just did it to save line space - could replace with the longhand everywhere if necessary, but could be painful to read.
Particularly the   
```
using TreeEdges = std::vector<std::vector<SmartEntry>>;
using Forest = std::vector<TreeEdges>;
```
which I did so I could write `vector<Forest>` rather than `vector<vector<vector<vector<SmartEntry>>>>` which seems ludicrous

Reply:
Ah, I didn't mean the using declaration, I meant avoiding the need to look up variables in a hash, rather than knowing where they are directly.